src/cargo/ops/vendor.rs

```diff
@@ -173,20 +168,28 @@ fn sync(
     let mut versions = HashMap::new();
     for id in ids.keys() {
         let map = versions.entry(id.name()).or_insert_with(BTreeMap::default);
-        if let Some(prev) = map.get(&id.version()) {
-            bail!(
-                "found duplicate version of package `{} v{}` \
-                 vendored from two sources:\n\
-                 \n\
-                 \tsource 1: {}\n\
-                 \tsource 2: {}",
-                id.name(),
-                id.version(),
-                prev,
-                id.source_id()
-            );
+        use std::collections::btree_map::Entry;
+        match map.entry(id.version()) {
+            Entry::Vacant(entry) => {
+                entry.insert(id.source_id());
+            }
+            Entry::Occupied(_) => {
+                // Detect any duplicate versions and handle it later.
+                contains_duplicate_versions |= true;
+            }
+        }
+    }
+
+    // The layout of the vendor directory depends on if there is any duplicate
+    // version. cargo-vendor auto switches between the two layouts and deletes
+    // the old vendor directory before vendoring crates.
+    match (sources_file.exists(), contains_duplicate_versions) {
```

Comment:
I'm a bit confused still on what's going on here. My current impression is that `cargo vendor` gets to blow away the entire `vendor` directory unless you pass `--no-delete`, and in that situation it simply just appends what's being added. Is that model still accurate?

If so, how come this is removing contents plus the block below handling `no_delete`?

Reply:
Sorry about I haven't explained it too much. This piece was copied from https://github.com/alexcrichton/cargo-vendor/blob/07570e23d0841936cea8af2a125484f4d10ccdfa/src/main.rs#L147-L152. 

The merged and no-merged layout are not compatible with each other, so cargo must delete something to make it work.

Reply:
Sorry about I haven't explained it too much. This piece was copied from https://github.com/alexcrichton/cargo-vendor/blob/07570e23d0841936cea8af2a125484f4d10ccdfa/src/main.rs#L147-L152. 

The merged and no-merged layout are not compatible with each other, so cargo must delete something to make it work. For example, we have a crate with `log` and `serde` vendored,

```
vendor/
├── log/
└── serde/
```

If we add the other `log` deps from git and vendor it, the directory tree would look like:

```
vendor/
├── registry-1ecc6299db9ec823/ # from crates.io
│  ├── log/
│  └── serde/
└── git-1936cea8af2a1111/ # from git
   └── serde/
```

Since both `registry-1ecc6299db9ec823` and `git-1936cea8af2a1111` are valid package names, we need to delete them all to avoid ambiguity or even errors.

Off the top of my head, one solution is that cargo can put crates from `crates.io` source at the first level, and duplicate version crate from other sources at the second level with folder names contains special characters which make them not valid package names. For instances, prefix sources with `~`:

```
vendor/
├── log/ # from crates.io
└── serde/ # from crates.io
└── ~git-1936cea8af2a1111/ # git source
   └── serde/
```

With this solution, the vendored crates becomes appendable and can respect `no-delete` again. Also, the use of `vendor/.sources` file are not necessary anymore.